### PR TITLE
Document deriving Functor, Foldable, Traversable

### PR DESCRIPTION
**Pull request by rhendric**
Adapted from @JordanMartinez's changelogs in purescript/purescript#4371; I think this level of explanation is better off here than in release notes. Not to be merged until the feature lands (either in that PR or in another).

### CODE REVIEW

**Comment by JordanMartinez** [comment]
Ah good point! That will clean up the release notes significantly.

File: language/Type-Classes.md [hunk 1]
```diff
@@ -183,8 +183,101 @@ Currently, instances for the following classes can be derived by the compiler:
...
+
+All three of these classes can be derived for certain data types, provided the following requirements are met:
+
+1. As these classes all have kind `(Type -> Type) -> Constraint`, the data type must have at least one type parameter.
+2. Every occurrence of the final parameter in the data type must be in a legal position. Legal positions are:
+   * Data constructor arguments
+   * A field in a record type that itself appears in a legal position
+   * The rightmost argument to a type constructor that itself appears in a legal position. In this case, the type constructor must have an instance of the same class being derived. (For example, function types returning the parameter can be used when deriving `Functor` but not when deriving `Foldable` or `Traversable`.)
+
+(Note for Haskell programmers: unlike the very similar feature in GHC, PureScript does not currently consider the `a` in `((a -> b) -> b)` to be a legal position in the above sense when deriving `Functor`.)
```
Comment (JordanMartinez) [inline language/Type-Classes.md:201]: Maybe this point should also be documented in the "difference from Haskell" page?

Comment (rhendric) [reply language/Type-Classes.md:201]: Good question, no strong opinions from me.

Comment (rhendric) [reply language/Type-Classes.md:201]: Done (the list of GHC extensions supported by PureScript needed updating anyway).

**Comment by JordanMartinez** [comment]
LGTM!

**Review by JordanMartinez — APPROVED** [review]

**Pull request merged**